### PR TITLE
[data,ci,docs] feat: add torchcodec-based video processing with ffmpeg support and comprehensive testing

### PR DESCRIPTION
**Pull request by TimYangst**
### What does this PR do?

> Replaces PyAV-based video processing with torchcodec for significant performance improvements (18.2x speedup on average, up to 19.4x for audio processing), refactors audio extraction utilities, adds comprehensive test suite with functional and benchmark tests, and updates documentation with installation instructions. This enhancement adds robust video processing capabilities with multi-format input support.

### Checklist Before Starting

- [ ] Search for similar PRs. Paste at least one query link here: https://github.com/ByteDance-Seed/VeOmni/pulls?q=is%3Apr+video+processing
- [x] Format the PR title as `[{modules}] {type}: {description}` (This will be checked by the CI)
  - Modules: `data`, `ci`, `docs`
  - Type: `feat`
  - Description: add torchcodec-based video processing with ffmpeg support and comprehensive testing

### Test

All 10 tests pass successfully with significant performance improvements:

**Functional Tests (7):**
```
tests/data/multimodal/test_video_utils.py::test_fetch_videos_from_path PASSED
tests/data/multimodal/test_video_utils.py::test_fetch_videos_from_bytes PASSED
tests/data/multimodal/test_video_utils.py::test_fetch_videos_from_pil_images PASSED
tests/data/multimodal/test_video_utils.py::test_fetch_videos_from_dict PASSED
tests/data/multimodal/test_video_utils.py::test_fetch_videos_without_audio PASSED
tests/data/multimodal/test_video_utils.py::test_fetch_videos_with_frame_constraints PASSED
tests/data/multimodal/test_video_utils.py::test_fetch_videos_multiple_inputs PASSED
```

**Benchmark Tests (3) - Performance Improvements:**

1. **Basic Video Processing** - 18.2x faster:
   - **Before:** 1491.42 ms average (std dev: 51.27 ms)
   - **After:** 81.96 ms average (std dev: 2.63 ms)
   - Video shape: torch.Size([5, 3, 546, 966])
   - Audio samples: 89,885
   - Total pixels processed: 2,637,180

2. **Multi-Resolution Processing** - 9.4-10.1x faster:
   ```
   Resolution         Frames    Shape           Avg Time (ms)    Std Dev    FPS
   Low (224x224)      11        546x966         158.18          7.42       69.54
   Medium (448x448)   11        546x966         172.60          18.02      63.73
   High (896x896)     11        672x1204        188.33          21.01      58.41
   ```
   - **Before:** 1486-1907 ms (std dev: 64.99-86.52 ms)
   - **After:** 158-188 ms (std dev: 7.42-21.01 ms)

3. **Audio Processing** - 17.1-19.4x faster:
   ```
   Config                  Avg Time (ms)    Std Dev    Audio Samples
   No Audio                77.00           5.99       0
   With Audio (16kHz)      86.75           3.64       89,885
   With Audio (24kHz)      85.57           3.62       134,827
   ```
   - **Before:** 1465-1478 ms (std dev: 48.88-89.65 ms)
   - **After:** 77-87 ms (std dev: 3.62-5.99 ms)

### API and Usage Example

No breaking API changes. Enhanced `fetch_videos()` to support multiple input formats:

```python
from veomni.data.multimodal.video_utils import fetch_videos

# 1. File path input
videos, audios = fetch_videos(["/path/to/video.mp4"], fps=2, sample_rate=16000)

# 2. Bytes input
with open("/path/to/video.mp4", "rb") as f:
    video_bytes = f.read()
videos, audios = fetch_videos([video_bytes], fps=2, sample_rate=16000)

# 3. PIL Images list (new)
from PIL import Image
images = [Image.open(f"frame_{i}.jpg") for i in range(10)]
videos, _ = fetch_videos([images], fps=2)

# 4. Dict with numpy arrays (new)
video_dict = {
    "video": np.array(...),  # Shape: (T, H, W, C)
    "audio": np.array(...),
    "video_fps": 30.0,
    "audio_fps": 16000
}
videos, audios = fetch_videos([video_dict], fps=2, sample_rate=16000)

# 5. Multiple mixed inputs (new)
videos, audios = fetch_videos(["/path/to/video.mp4", video_bytes, images], fps=2)
```

**Installation Requirements:**
```bash
# Install VeOmni with audio dependencies
uv sync --locked --extra gpu --extra audio

# Install ffmpeg (required for torchcodec)
sudo apt-get install ffmpeg  # Ubuntu/Debian
brew install ffmpeg          # macOS
```

### Design & Code Changes

**1. Core Video Processing Engine ([veomni/data/multimodal/video_utils.py](veomni/data/multimodal/video_utils.py))**

- **Replaced PyAV with torchcodec:** Switched from PyAV's `av.open()` to torchcodec's `VideoDecoder` for 18.2x average performance improvement (up to 19.4x for audio processing)
- **Added ffmpeg availability check:** `is_ffmpeg_available()` function in [veomni/utils/import_utils.py](veomni/utils/import_utils.py) to validate torchcodec dependencies
- **Multi-format input support:**
  - PIL Image lists: Direct frame-to-tensor conversion for image sequences
  - Dict format: Support for pre-loaded numpy arrays with metadata
  - Backward compatible: Maintains support for file paths and bytes
- **Improved frame sampling:** Enhanced `calculate_frame_indices()` with better edge case handling and clearer logic
- **Optimized video decoding:** Uses torchcodec's efficient batch frame extraction with proper duration alignment

**2. Audio Extraction Utilities ([veomni/data/multimodal/audio_utils.py](veomni/data/multimodal/audio_utils.py))**

- **New `extract_audio_from_video()` function:** Extracts audio from video using PyAV with OOM protection
- **Duration-based safety limits:** Prevents memory overflow with configurable max duration
- **Graceful degradation:** Returns None for videos without audio streams
- **Multi-channel support:** Automatic conversion to mono audio

**3. Import Utilities ([veomni/utils/import_utils.py](veomni/utils/import_utils.py))**

- **New `is_ffmpeg_available()` function:** Checks ffmpeg availability for torchcodec dependency validation

**4. Comprehensive Test Suite ([tests/data/multimodal/test_video_utils.py](tests/data/multimodal/test_video_utils.py))**

- **7 Functional Tests:**
  - `test_fetch_videos_from_path`: File path validation
  - `test_fetch_videos_from_bytes`: Bytes input validation
  - `test_fetch_videos_from_pil_images`: PIL Image list support (new)
  - `test_fetch_videos_from_dict`: Dict format support (new)
  - `test_fetch_videos_without_audio`: Audio-disabled mode
  - `test_fetch_videos_with_frame_constraints`: Frame count constraints
  - `test_fetch_videos_multiple_inputs`: Mixed input types (new)

- **3 Benchmark Tests:**
  - `test_benchmark_fetch_videos_from_path`: Basic performance metrics (5 iterations)
  - `test_benchmark_fetch_videos_different_resolutions`: Multi-resolution performance
  - `test_benchmark_audio_processing`: Audio resampling performance
  - All tests skip gracefully when ffmpeg is unavailable

**5. Dependency Management**

- **pyproject.toml:**
  - Added `torchcodec>=0.6.0,<0.8.0` for GPU environments
  - Added `torchcodec>=0.4.0,<0.6.0` for NPU environments
  - Added torchcodec to `audio` extra dependencies
  - Created `video` extra as an alias for audio/video processing dependencies
  - Added pytest markers for benchmark tests
  - Updated override-dependencies for proper version resolution

- **uv.lock:** Updated with torchcodec 0.5-0.7.0 across platforms

**6. CI/CD Updates**

- **.github/workflows/npu_unit_tests.yml:** Added ffmpeg installation step and `--extra audio` in dependency sync
- **Test markers:** All video tests marked with `@pytest.mark.skipif(not is_ffmpeg_available())`

**7. Documentation ([docs/get_started/installation/](docs/get_started/installation/))**

- **install.md:** Added ffmpeg installation instructions for GPU environments
- **install_ascend.md:** Added ffmpeg installation instructions for NPU/Ascend environments
- **Usage notes:** Clarified `--extra audio` dependency for video processing

**Key Performance Metrics:**
- **18.2x faster** video decoding (1491ms → 82ms for basic processing)
- **9.4-10.1x faster** multi-resolution processing (1486-1907ms → 158-188ms)
- **17.1-19.4x faster** audio processing (1465-1478ms → 77-87ms)
- **Significantly lower variance:** Standard deviation reduced from 48-90ms to 2.63-21.01ms (up to 34x more stable)

### Checklist Before Submitting

> [!IMPORTANT]
> Please check all the following items before requesting a review, otherwise the reviewer might deprioritize this PR for review.

- [ ] Read the [Contribute Guide](https://github.com/ByteDance-Seed/VeOmni/blob/main/CONTRIBUTING.md).
- [ ] Apply [pre-commit checks](https://github.com/ByteDance-Seed/VeOmni/blob/main/CONTRIBUTING.md?plain=1#L22): `make commit && make style && make quality`
- [ ] Add / Update [the documentation](https://github.com/ByteDance-Seed/VeOmni/blob/main/docs). ✓ Documentation updated in install.md and install_ascend.md
- [x] Add unit or end-to-end test(s) to [the CI workflow](https://github.com/ByteDance-Seed/VeOmni/tree/main/.github/workflows) to cover all the code. ✓ Added 10 comprehensive tests (7 functional + 3 benchmark)

### CODE REVIEW

File: veomni/data/multimodal/video_utils.py [hunk 1]
```diff
@@ -16,6 +20,28 @@
...
+
+    TORCHCODEC_AVAILABLE = True
+except ImportError:
+    TORCHCODEC_AVAILABLE = False
+    VideoDecoder = None
+
+_FFMPEG_AVAILABLE = None
+
+
+def is_ffmpeg_available():
```
Comment (Luosuu) [inline veomni/data/multimodal/video_utils.py]: do we have version limit for ffmpeg

File: veomni/data/multimodal/video_utils.py [hunk 2]
```diff
@@ -130,89 +162,248 @@ def smart_resize(
...
+    """
+    Convert dict to video tensor and audio array.
+    Expected keys: 'video' (required), 'audio' (optional), 'video_fps' (optional), 'audio_fps' (optional).
+    """
+    if "video" not in video_dict:
+        raise ValueError("Dict input must contain 'video' key")
+
+    video_np = video_dict["video"]
+    # Assume video_np is (T, H, W, C) or (T, C, H, W)
+    if video_np.ndim == 4:
```
Comment (Luosuu) [inline veomni/data/multimodal/video_utils.py:280]: maybe add some logs here? it seems easy to get wrong

File: veomni/data/multimodal/video_utils.py
Comment (Luosuu) [inline veomni/data/multimodal/video_utils.py]: tbh, I am confused with prefix `smart_*`..

File: veomni/data/multimodal/video_utils.py (hunk 1 above)
Comment (TimYangst) [reply veomni/data/multimodal/video_utils.py]: Add a link there, from the document of torchcodec https://github.com/meta-pytorch/torchcodec, a widely version of ffmpeg[4, 8] is supported.

File: veomni/data/multimodal/video_utils.py (hunk 2 above)
Comment (TimYangst) [reply veomni/data/multimodal/video_utils.py:280]: done

File: veomni/data/multimodal/video_utils.py
Comment (TimYangst) [reply veomni/data/multimodal/video_utils.py]: Add more comments to the methods. The smart_ method name are followed with the previous version. 
Basically: 
 The "smart" aspect refers to the intelligent handling of multiple conflicting constraints:
    1. Target FPS sampling (fps parameter)
    2. Frame count bounds (min_frames, max_frames)
    3. Frame factor alignment (for models requiring specific frame counts like multiples of 8/16)
    4. Automatic padding when needed (duplicates last frame if target > available frames)
    5. Ensures at least 1 frame output regardless of constraints
    
For audio:
Smart audio resampling to target sample rate.

    The "smart" aspect refers to:
    1. Automatic resampling to standard sample rates (e.g., 16kHz for speech models)
    2. Handles None/missing audio gracefully (returns None without error)
    3. Uses high-quality librosa resampling with anti-aliasing

Comment (Luosuu) [reply veomni/data/multimodal/video_utils.py]: yeah maybe we should have more comments here

Comment (TimYangst) [reply veomni/data/multimodal/video_utils.py]: done

File: veomni/data/multimodal/video_utils.py [hunk 3]
```diff
@@ -1,21 +1,39 @@
...
 
 from ...utils import logging
 
 
 logger = logging.get_logger(__name__)
 
+_FFMPEG_AVAILABLE = None
+
+
+def is_ffmpeg_available():
```
Comment (Coach257) [inline veomni/data/multimodal/video_utils.py]: move this to veomni.utils.import_utils

File: veomni/data/multimodal/video_utils.py [hunk 4]
```diff
@@ -130,89 +246,202 @@ def smart_resize(
...
+    if pad_count > 0:
+        last_frame = resized_frames[-1:].expand(pad_count, -1, -1, -1)
+        final_frames = torch.cat([resized_frames, last_frame], dim=0)
     else:
-        raise NotImplementedError
+        final_frames = resized_frames
+
+    # Audio extraction with PyAV
+    audio, audio_fps = None, None
+    if use_audio_in_video:
```
Comment (Coach257) [inline veomni/data/multimodal/video_utils.py:386]: maybe move these `audio_reading_process` to audio_utils

File: veomni/data/multimodal/video_utils.py (hunk 3 above)
Comment (TimYangst) [reply veomni/data/multimodal/video_utils.py]: Done

File: veomni/data/multimodal/video_utils.py (hunk 4 above)
Comment (TimYangst) [reply veomni/data/multimodal/video_utils.py:386]: Done

**Review by Coach257 — APPROVED** [review]
